### ISSUE TITLE
Bug: manual slope selection plot does not update if selected point is before current slope

### ISSUE THREAD

**Issue by js3110** [pharmaverse/aNCA#121]
### Description
When a manual slope selection is added, the plot only updates if the point selected is > time than the previous selection, otherwise it doesn't update visually. In the NCA results and in the table everything is fine, it is just the visuals of the plot.

### To reproduce
Run NCA -> Slope selector -> select a point that is before the current slope (black) and then click the end point. The black point selected does not turn green.

### Expected behaviour
The black point selected should turn green and be included in the line of best fit

**Comment by m-kolomanski** [comment]
Hi @js3110 , @Gero1999 ,

I have investigated the issue briefly, the dot on the plot is turned green based on the following:
```R
IX %in% lambda_z_ix_rows$IX ~ "hl.included",
```
and in order to pass that, it needs to go through the following checks:
```R
lambda_z_ix_rows <- conc_pknca_df %>%
    filter(
      DOSNO == dosno,
      USUBJID == usubjid,
      !exclude_half.life,
      TIME >= sum(
        subset(
          lambda_res,
          lambda_res$PPTESTCD == "lambda.z.time.first",
          select = c("start", "PPORRES")
        )
      )
    ) %>%
    arrange(IX) %>%
    slice(0:lambda_z_n_points)
```
and the cases we are talking about do not pass the `TIME >= lamba.z.time.first` check, as @Gero1999 suggested. However, even if we remove this check, the results are still wrong because of the slicing at the end, which always takes the same number of points based on:
```R
lambda_z_n_points <- as.numeric(lambda_res$PPORRES[lambda_res$PPTESTCD == "lambda.z.n.points"])
```
The `lambda.z.n.points` in the results for a given subject does not seem to change based on slope selector. I am afraid my knowledge of PKNCA package and NCA analysis in general is too low to allow me to comment on why that my be and if that is the correct behavior - do you have any thoughts on that?

Removing both time check and slicing seems to make the plots work as expected, however all dots are green at the beginning. I do not know if that is what we would want and whether is has any implications to the data / analysis?

**Comment by js3110** [comment]
@m-kolomanski  removing the checks and having all dots green is definitely not an option.
I think that if the previous point is added to the selection then we would want `lambda.z.n.points` and time first to update to include the new point, same as how n.points should update if for example IX 5 is selected on a plot that used to include IX 6.

The way the NCA works is it takes the points included (in this situation the ones in green) and calculates parameters based off of these (area under curve, regression of line etc). The PKNCA package automatically calculates it based on best fit, but the User needs to be able to manually choose the line incase the best fit does not meet their needs/ expectations.

**Comment by Gero1999** [comment]
Hello @m-kolomanski thanks for pointing this out! Indeed, this is then the conflict between PKNCA and the function. Here is a connected issue indeed #153. You are right, one solution to prevent these conflicts would be that whenever there is a manual selection all points corresponding to that patient/profile are processed independently with different `PKNCA.options()`. I can try to propose an easy fix

**Comment by m-kolomanski** [comment]
@Gero1999 Thank you for investigating that thoroughly!

I think with the points 1. and 2., we do have a bit of an unclear data flow in the application. I think that we should focus on modularizing and refactoring the NCA tab (#144) and the think about the exact data flow we want to achieve, what updates what and when etc, so we know exactly what is going on with the data. Then we will be able to go back to this issue, and fix the module  so that the analysis pipeline actually considers user inputs.

**Comment by m-kolomanski** [comment]
Hi @js3110 @Gero1999,

I was thinking how to improve this situation and I would like your opinion, since this changes the workflow a bit.

So currently the `slope_selector` module has an **Apply** button that _in principle_ applies the filters to `mydata()` object and triggers NCA recalculation. I would like to suggest **removing** this button. Instead, when the first changes to the selection is made, the application should display a warning, something along the lines:
> Settings changes detected, please re-run the analysis to update the results.
and then the filters get applied when the user clicks **Run NCA**. 

We could also implement the same kind of warning to other available settings in the Setup tab. This would also centralize the required actions (does not matter what changes have been made, user always has to click **Run NCA** button to update).

Aaaand we improve data flow, since `slope_selector` does not need to trigger anything, does not have to update any data, it just returns a list of rules that then is applied when NCA runs.

What do you think?

**Comment by js3110** [comment]
@m-kolomanski I think that is a good idea and probably the best way to move forward to avoid a too complex workflow. However, I would suggest waiting until #177 has been approved and merged, or at least using it as the parent branch, as I have made quite a few changes to the slope selectors so this would avoid having too many conflicts!

**Comment by m-kolomanski** [comment]
@js3110 All clear. We can merge #177 first, then I can deal with the conflicts. The PR is marked as a draft and checks are failing, but reviews are requested - is it ready for review, what is the status?

**Comment by js3110** [comment]
Hi @m-kolomanski sorry I forgot to change it from a draft to ready. It is ready for review, the checks that are failing are small things so I am currently trying to fix (lint, spellcheck etc), so it shouldn't impact the review. I will try to have it passing everything in the next hour!

**Comment by Gero1999** [comment]
Hello @m-kolomanski I understand that the idea could seem better but I might disagree in this one with @js3110! There are 2 problematic points from my perspective:

1) The slope selector will be at some point be done pre NCA settings (thus also before running the NCA). Currently, this isn't possible due to the workflow's structure and dependencies. However, after addressing issue #174 , we can make mydata() dynamic and create a pre_res_nca object to run NCA on the necessary parameters for the slope selector.

2) Our stakeholders prefer an "Apply" or "Save" button to ensure changes aren't applied to final results without explicit confirmation. They like to experiment with selections and statistics but often don't apply changes due to the need for a REASON, especially for Type="Selection". Without a save button, they might miss this step when rerunning the NCA.

In summary, while I agree on improving and changing the module's interaction with the rest of the code workflow, I believe keeping the "Apply" or "Save" button is essential. And then that at some point the `slope selector` tab needs to go before `NCA settings`.

**Comment by js3110** [comment]
> Hello [@m-kolomanski](https://github.com/m-kolomanski) I understand that the idea could seem better but I might disagree in this one with [@js3110](https://github.com/js3110)! There are 2 problematic points from my perspective:
> 
> 1. The slope selector will be at some point be done pre NCA settings (thus also before running the NCA). Currently, this isn't possible due to the workflow's structure and dependencies. However, after addressing issue [Enhancement: NCA parameter input #174](https://github.com/pharmaverse/aNCA/pull/174) , we can make mydata() dynamic and create a pre_res_nca object to run NCA on the necessary parameters for the slope selector.
> 2. Our stakeholders prefer an "Apply" or "Save" button to ensure changes aren't applied to final results without explicit confirmation. They like to experiment with selections and statistics but often don't apply changes due to the need for a REASON, especially for Type="Selection". Without a save button, they might miss this step when rerunning the NCA.
> 
> In summary, while I agree on improving and changing the module's interaction with the rest of the code workflow, I believe keeping the "Apply" or "Save" button is essential. And then that at some point the `slope selector` tab needs to go before `NCA settings`.

point 1 makes sense @Gero1999, however it wouldn't make a difference if the apply button is removed as either way the ápply' or 'run nca' would rerun the NCA. At the moment they both trigger the same workflow. As long as the plots can update before rerunning the nca, this shouldn't matter. And regarding your second point, if they have to click Run NCA again then this is the same as clicking apply. It is actually the same method as they are using in the current software so I don't think it will be an issue for the Users.

**Comment by m-kolomanski** [comment]
@js3110 @Gero1999,

The slope plots would work as they do now - upon each selection / exclusion they refresh. As @js3110 pointed out, currently the `Apply` button and `Run NCA` do basically the same thing. The user will still be able to play freely with the selections/exclusions and see the results in the plots. @Gero1999, the explicit confirmation would still be there, just that this would be done by clicking `Run NCA` instead of `Apply`.

**Comment by Gero1999** [comment]
We can go for it and try, but just to explain better for me the logical workflow for the user would be: 
1) Map dataset 
2) Checks/modifies the slope plots
3) Confirm changes saving them in `mydata`
4) Defines the NCA settings 
5) run NCA. For me that is the logical workflow for the user

**Comment by m-kolomanski** [comment]
Thank you for spelling it out like that, it does make a lot of sense from the users perspective.

Programatically, I do not think it matters if we apply the changes at any arbitrary time when user presses the button or just before the analysis, all at once.

I would propose keeping it simple and with line with the current capabilities of the module.

Then, if a) we receive user feedback that the workflow is unintuitive or b) it makes sense to add an apply button in the context of some new capabilities (like running NCA for the purpose of updating slope selector), then we can add it back in in a more thought-out manner, in line with the goal it should achieve.

**Comment by Gero1999** [comment]
Makes sense to me @m-kolomanski thanks!

**Comment by js3110** [comment]
@m-kolomanski @Gero1999 should we rewrite this issue to account for the additional changes we plan to make, or make a new issue that is separate to this bug fix?

**Comment by m-kolomanski** [comment]
I am always in favor of splitting the tasks into separate issues wherever possible, especially since we do struggle with overscoping.

Even if the change is 1 line long, I would still go for separate tasks as long as it can be formulated as a separate and cohesive issue / bug, since that means we can implement -> review -> merge -> forget very swiftly.

**Comment by js3110** [comment]
This issue is now unblocked, and high priority

**Comment by m-kolomanski** [comment]
Hey @js3110 @Gero1999, can we reiterate what is the desired solution for this problem?

As far as I remember, we do not want to _hack in_ the green dots based on manual slopes selection. Instead, we wanted to make sure that the underlying data objects are up-to-date and provide the desired result.

Slope plots are now based on their own copy of pknca data with proper setup, and rely on their own nca results.

However, this does not impact how the plots ale colored.

Going back to my previous comment:

_The dot on the plot is turned green based on the following:__
```R
IX %in% lambda_z_ix_rows$IX ~ "hl.included",
```
_and in order to pass that, it needs to go through the following checks:_
```R
lambda_z_ix_rows <- conc_pknca_df %>%
    filter(
      DOSNO == dosno,
      USUBJID == usubjid,
      !exclude_half.life,
      TIME >= sum(
        subset(
          lambda_res,
          lambda_res$PPTESTCD == "lambda.z.time.first",
          select = c("start", "PPORRES")
        )
      )
    ) %>%
    arrange(IX) %>%
    slice(0:lambda_z_n_points)
```
_and the cases we are talking about do not pass the `TIME >= lamba.z.time.first` check. However, even if we remove this check, the results are still wrong because of the slicing at the end, which always takes the same number of points based on:_
```R
lambda_z_n_points <- as.numeric(lambda_res$PPORRES[lambda_res$PPTESTCD == "lambda.z.n.points"])
```

So the dot coloring hinges on `lambda.z.time.first` and `lambda.z.n.points` from the result. I have added filtering based on current selections to the results object:
```R
        result_obj <- pknca_data() %>%
          filter_slopes(
            manual_slopes(), profiles_per_patient(), slopes_groups()
          ) %>%
          PKNCA_calculate_nca()
```

however, **changing slopes selection does not change the above variables**. I can still break it by selecting too few points (then the values are set to `NA`), but if I select a wide range of points (e.g. 1:7), those values stay the same in the results meaning the dots are also colored the same.

To be honest I am stuck and I don't know how achieve the desired result, without directly imposing the `manual_slopes()` selection on the plot. Any advice would be appreciated.

**Comment by js3110** [comment]
Hi @m-kolomanski 

I have a feeling that the problem may go a bit deeper. Since the lambda.z.time.first SHOULD be changing if an inclusion is added (the new point should become the time first). So this could be an issue with the calculation of this part (using PKNCA package) rather than an issue with the slopes plots. 

If you prefer, I can take over this issue and investigate this side a bit further?

**Comment by m-kolomanski** [comment]
@js3110 That would be lovely, since I am not sure where to start really!

**Comment by js3110** [comment]
@m-kolomanski From what I can see, the problem is due to the way that we are calculating 'inclusions' in the app.
In the PKNCA package there are two options for manual half life calculation: `exclude_half.life` and `include_half.life`. For exclude its pretty straightforward- just set FALSE for all except for the points you want to exclude.

However, for include, from what I can see, currently the only option is to set TRUE for the points you want, and FALSE for the points you don't want, but if you set it as NA (as in just continue with the automatic curve stripping calculation),` pk.nca()` doesn't work and gives you errors. This feature seems to only work if you want to do one subject at a time, which unfortunately is not what we are after.

In our app, we currently do this:

```
mutate(exclude_half.life = {
      if (any(is.included.hl)) {
        is.excluded.hl | !is.included.hl
      } else {
        is.excluded.hl
      }
    })
```

which is fine for exclusions, but for the inclusions only works if the points were already included in the automatic selection- essentially instead of 'including' points, we are just excluding the ones outside of our selection. Which i assume was a temporary workaround since include_half.life doesn't allow for what we are looking for, @Gero1999 ?

From what I can see in the PKNCA github, the issue [387](https://github.com/billdenney/pknca/issues/387) is open, which seems to be related to what we are after. Maybe @billdenney could provide some insight on if this will solve our issue?

Otherwise, an alternative solution that I could think of is for subjects where inclusions are added, we filter them from the main results and then calculate a separate pk.nca() for each individual subject, using the inclusions. Not ideal but would get the job done...

**Comment by Gero1999** [comment]
Hey Jana, indeed back on the first time I developed the functions I was forced to make a workaround to allow the plots logic of selection (be capable of making simultaneously inclusions and exclusions of the same points). Even if  we have columns called `exclude_half.life` and `include_half.life`, the truth is that these two are transformed in one new column that excludes the points via PKNCA. So I did not intend to use the "inclusion" action of PKNCA

**Comment by Gero1999** [comment]
So in short, when PKNCA was given a point with both half.life included and excluded, it gave priority to the inclusions. This for plotting/selection is not what we want. So I made that workaround to just use the exclusion functionality

**Comment by js3110** [comment]
> So in short, when PKNCA was given a point with both half.life included and excluded, it gave priority to the inclusions. This for plotting/selection is not what we want. So I made that workaround to just use the exclusion functionality

but were you able to make include_half.life work for multiple subjects? for me it gives an error if I set it to NA

**Comment by billdenney** [comment]
@js3110, The development version of PKNCA is more forgiving about the include and exclude half-life columns where `NA` is treated as `FALSE`. I think that may resolve what you're looking for about those columns.

The PR related to billdenney/pknca#387 will provide a method to extract which points were used for half-life calculation. I see that as helping with this but not being integral to direct points selection for inclusion/exclusion.

The behavior about giving inclusion priority over exclusion is intentional based on the way that they both work. With inclusion, you are selecting all the points to include. PKNCA does not make any choices. With exclusion, you are selecting points to exclude, and PKNCA makes choices among the remaining points. It is a good point that combined inclusion/exclusion is not likely what people want, so that should likely be an error in PKNCA.

**Comment by billdenney** [comment]
Good question. If the column is all `NA` for an interval, then it is ignored. If any values are `TRUE` or `FALSE` in an interval, then the `NA` are considered as `FALSE`.

If that's not clear in the documentation, can you please raise an issue in PKNCA?
